### PR TITLE
Add SVM footprint keywords to product headers

### PR DESCRIPTION
**Pull request by stsci-hack**
Addresses #537.

### CODE REVIEW

File: drizzlepac/hlautils/product.py [hunk 1]
```diff
@@ -23,6 +24,14 @@
 log = logutil.create_logger(__name__, level=logutil.logging.NOTSET, stream=sys.stdout,
                             format=SPLUNK_MSG_FORMAT, datefmt=MSG_DATEFMT)
 
+# Define keywords to be added to SVM products to describe the overlap of
+# exposures onto the output WCS footprint as determined from the SkyFootprint masks
+MASK_KWS = {"NPIXFRAC": [None, "Fraction of pixels with data"],
+            "MEANEXPT": [None, "Mean exposure time per pixel with data"],
+            "MEDEXPT": [None, "Median exposure time per pixel with data"],
+            "MEANNEXP": [None, "Mean exposures per pixel with data"],
+            "MEDNEXP": [None, "Median exposures per pixel with data"],
```
Comment (mdlpstsci) [inline drizzlepac/hlautils/product.py]: Clarify the commentary part of the keyword for MEANNEXP and MEDNEXP - "[Mean|Median] **_number of_** exposures..."

Comment (stsci-hack) [reply drizzlepac/hlautils/product.py]: Fixed in b421222.

File: drizzlepac/hlautils/product.py [hunk 2]
```diff
@@ -154,9 +169,17 @@ def generate_footprint_mask(self):
         """
         footprint = cell_utils.SkyFootprint(self.meta_wcs)
         exposure_names = [element.full_filename for element in self.edp_list]
-        footprint.build(exposure_names)
+        footprint.build(exposure_names, scale=True)
         self.mask = footprint.total_mask
 
+        # Compute footprint-based SVM-specific keywords for product image header
+        good_pixels = self.mask > 0
+        self.mask_kws['NPIXFRAC'][0] = good_pixels.sum()
```
Comment (mdlpstsci) [inline drizzlepac/hlautils/product.py]: This looks to be the sum of pixels with good data versus a fraction (good_pixels/all_pixels).

File: drizzlepac/hlautils/product.py [hunk 3]
```diff
@@ -272,10 +303,35 @@ def wcs_drizzle_product(self, meta_wcs):
...
+        self.mask = footprint.total_mask
+
+        # Compute footprint-based SVM-specific keywords for product image header
+        good_pixels = self.mask > 0
+        self.mask_kws['NPIXFRAC'][0] = good_pixels.sum()
+        self.mask_kws['MEANEXPT'][0] = np.mean(footprint.scaled_mask[good_pixels])
+        self.mask_kws['MEDEXPT'][0] = np.median(footprint.scaled_mask[good_pixels])
+        self.mask_kws['MEANNEXP'][0] = np.mean(self.mask[good_pixels])
+        self.mask_kws['MEDNEXP'][0] = np.median(self.mask[good_pixels])
+
```
Comment (mdlpstsci) [inline drizzlepac/hlautils/product.py]: It looks like _generate_footprint_mask()_  and _wcs_drizzle_product()_ are basically the same for the _TotalProduct_ and _FIlterProduct_ classes.  It is clear you do not want the _ExposureProduct_ class to have these methods.  You could put these method in the _HAPProduct_ base class and then override the methods in the _ExposureProduct_ class with "pass" which is a cheapo solution. Alternatively, you could do what is discussed at this URL which I found compelling:
https://medium.com/@maouu/sorry-but-youre-wrong-aea1b88ffc03

Comment (stsci-hack) [reply drizzlepac/hlautils/product.py]: The alternative definitions using __getattribute__ and __dict__ for ExposureProduct was implemented successfully(?) in 871bdef.  Thanks for the pointer!

File: drizzlepac/hlautils/product.py (hunk 2 above)
Comment (stsci-hack) [reply drizzlepac/hlautils/product.py]: Also fixed in 871bdef.

**Pull request merged**